pkg/controller/job/job_controller.go

```diff
@@ -424,34 +457,37 @@ func (jm *JobController) syncJob(key string) error {
 		now := metav1.Now()
 		job.Status.StartTime = &now
 	}
-	// if job was finished previously, we don't want to redo the termination
-	if IsJobFinished(&job) {
-		return nil
-	}
 
+	jobFailed := false
 	var manageJobErr error
-	if pastActiveDeadline(&job) {
+	var newFailureConditionFunc func(record.EventRecorder) batch.JobCondition
+
+	jobHaveNewFailure := failed > job.Status.Failed
+
+	// check if the number of failed jobs increaded since the last syncJob
+	if jobHaveNewFailure && jobToManyFailed(previousRetry+1, job.Spec.BackoffLimit) {
+		jobFailed = true
+		newFailureConditionFunc = func(r record.EventRecorder) batch.JobCondition {
+			r.Event(&job, v1.EventTypeNormal, "BackoffLimitExceeded", "Job has reach the specified backoff limit")
+			return newCondition(batch.JobFailed, "BackoffLimitExceeded", "Job has reach the specified backoff limit")
+		}
+	} else if pastActiveDeadline(&job) {
 		// TODO: below code should be replaced with pod termination resulting in
 		// pod failures, rather than killing pods. Unfortunately none such solution
 		// exists ATM. There's an open discussion in the topic in
 		// https://github.com/kubernetes/kubernetes/issues/14602 which might give
 		// some sort of solution to above problem.
 		// kill remaining active pods
-		wait := sync.WaitGroup{}
-		errCh := make(chan error, int(active))
-		wait.Add(int(active))
-		for i := int32(0); i < active; i++ {
-			go func(ix int32) {
-				defer wait.Done()
-				if err := jm.podControl.DeletePod(job.Namespace, activePods[ix].Name, &job); err != nil {
-					defer utilruntime.HandleError(err)
-					glog.V(2).Infof("Failed to delete %v, job %q/%q deadline exceeded", activePods[ix].Name, job.Namespace, job.Name)
-					errCh <- err
-				}
-			}(i)
+		jobFailed = true
+		newFailureConditionFunc = func(r record.EventRecorder) batch.JobCondition {
+			r.Event(&job, v1.EventTypeNormal, "DeadlineExceeded", "Job was active longer than specified deadline")
```

Comment:
Should be `EventTypeWarning` not `EventTypeNormal`.

Reply:
Don't create methods like this, just record event and save condition in tmp variable.